symengine/series_generic.h

```diff
@@ -6,64 +6,76 @@
 #ifndef SYMENGINE_SERIES_GENERIC_H
 #define SYMENGINE_SERIES_GENERIC_H
 
-#include <list>
-
 #include <symengine/polynomial.h>
+#include <symengine/rational.h>
+#include <symengine/series.h>
 
 namespace SymEngine {
 //! UnivariateSeries Class
-class UnivariateSeries : public Basic {
-public:
-    //! `var_` : Variable of the UnivariateSeries
-    //! `poly_` : holds the UnivariatePolynomial of the series
-    //! `prec_` : precision of the UnivariateSeries, i.e. self = poly + O(var^prec)
+class UnivariateSeries : public SeriesBase<UnivariateExprPolynomial, Expression, UnivariateSeries> {
     // UnivariateSeries 1 + 2*x + x**2 + O(x**5) has dict_ = {{0, 1}, {1, 2}, {2, 1}} with var_ = "x" and prec_ = 5
-    RCP<const Symbol> var_;
-    RCP<const UnivariatePolynomial> poly_;
     unsigned int prec_;
 public:
     IMPLEMENT_TYPEID(UNIVARIATESERIES)
-    //! Constructor of UnivariateSeries class
-    UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const RCP<const UnivariatePolynomial> &poly);
+    UnivariateSeries(const UnivariateExprPolynomial sp, const std::string varname, const unsigned degree) : SeriesBase(std::move(sp), varname, degree) {}
+    UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const UnivariateExprPolynomial poly);
     UnivariateSeries(const RCP<const Symbol> &var, const unsigned int& precision, const unsigned int& max_exp, map_uint_mpz&& dict);
     UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const map_uint_mpz &dict);
     //! Constructor using a dense vector of mpz_class coefficients
     UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const std::vector<mpz_class> &v);
-
-    static RCP<const UnivariateSeries> create(const RCP<const Symbol> &var,
-            const unsigned int &prec, const std::vector<mpz_class> &v) {
+    static RCP<const UnivariateSeries> create(const RCP<const Symbol> &var, const unsigned int &prec, const std::vector<mpz_class> &v) {
         return make_rcp<const UnivariateSeries>(var, prec, v);
     }
+      
+    static RCP<const UnivariateSeries> series(const RCP<const Basic> &t, const std::string &x, unsigned int prec);
+    virtual std::size_t __hash__() const;
+    virtual int compare(const Basic &o) const;
+    virtual RCP<const Basic> as_basic() const;
+    virtual umap_int_basic as_dict() const;
+    virtual RCP<const Basic> get_coeff(int) const;
+    static UnivariateExprPolynomial var(const std::string &s);
+      
+    static Expression convert(const Integer &x);
+    static Expression convert(const mpq_class &x);
+    static Expression convert(const Rational &x);
+    static Expression convert(const Number &x);
```

Comment:
Since the coefficient type is Expression, you don't need all 4 types. See https://github.com/symengine/symengine/blob/master/symengine/series_piranha.h#L61

Just the last one is enough